cardano-api/src/Cardano/Api/Block.hs

```diff
@@ -175,6 +213,33 @@ instance HasTypeProxy BlockHeader where
     data AsType BlockHeader = AsBlockHeader
     proxyToAsType _ = AsBlockHeader
 
+getBlockHeader :: forall era . Block era -> BlockHeader
+getBlockHeader (ShelleyBlock shelleyEra block) = case shelleyEra of
+  ShelleyBasedEraShelley -> go
+  ShelleyBasedEraAllegra -> go
+  ShelleyBasedEraMary -> go
```

Comment:
Is a case with the same code in each branch normally how we get relevant instances, or do we have some smarter way of doing this, e.g. [pattern synonyms can provide instances](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#typing-of-pattern-synonyms).